### PR TITLE
refactor: use pipenv to manage dependencies

### PR DESCRIPTION
**Pull request by 4141done**
# What
Using pip and `requirements.txt` led to issues where dependent library versions were
unpredictable and led to errors in some deployment environments.

This will also allow us to use semantic version properly without needing to explicitly
pin dependencies at certain versions but perform updates to libraries intentionally.

### Proposed changes
1. Install and use `Pipenv` which will generate a `Pipfile.lock` detailing the exact versions of dependencies and their dependencies.  This will help us have more reproducible builds since each dependency and subdependency will be installed at the exact version it was tested with.
2. Changed the code that determines the pulumi cli version to use the output of `pip` rather than reading from the requirements file (since that's not an exact version).

### Pipenv Usage
`pipenv` is compatible with any commands that can be given to `pip`.  There are some niceties that are good to know too:

#### graph
Prints the dependency graph with the `Pipfile` specs and the actual version resolved to.
```
$ pipenv graph
awscli==1.22.101
  - botocore [required: ==1.24.46, installed: 1.24.46]
    - jmespath [required: >=0.7.1,<2.0.0, installed: 1.0.0]
    - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.2]
      - six [required: >=1.5, installed: 1.16.0]
    - urllib3 [required: >=1.25.4,<1.27, installed: 1.26.9]
  - colorama [required: >=0.2.5,<0.4.4, installed: 0.4.3]
  - docutils [required: >=0.10,<0.16, installed: 0.15.2]
  - PyYAML [required: >=3.10,<5.5, installed: 5.4.1]
  - rsa [required: >=3.1.2,<4.8, installed: 4.7.2]
    - pyasn1 [required: >=0.1.3, installed: 0.4.8]
  - s3transfer [required: >=0.5.0,<0.6.0, installed: 0.5.2]
    - botocore [required: >=1.12.36,<2.0a.0, installed: 1.24.46]
      - jmespath [required: >=0.7.1,<2.0.0, installed: 1.0.0]
      - python-dateutil [required: >=2.1,<3.0.0, installed: 2.8.2]
        - six [required: >=1.5, installed: 1.16.0]
      - urllib3 [required: >=1.25.4,<1.27, installed: 1.26.9]
fart==0.1.5
  - pyperclip [required: Any, installed: 1.8.2]
kic-pulumi-utils==1.0.0.post37+git.fb325311.dirty
  - GitPython [required: >=3.1.18,<3.2.0, installed: 3.1.27]
    - gitdb [required: >=4.0.1,<5, installed: 4.0.9]
      - smmap [required: >=3.0.1,<6, installed: 5.0.0]
  - passlib [required: >=1.7.4,<2.0.0, installed: 1.7.4]
  - pyyaml [required: >=5.3.1,<6.0, installed: 5.4.1]

... output truncated
```

#### Figuring out updates
You can use `pipenv update --outdated` to display a list of packages that have newer versions in the upstream.
From there, you can choose to update them one by one with `pipenv update <package name>` which will update the package according to the rules specified in the `Pipfile`.  You may need to change the specification in the `Pipfile` to get the latest version.

https://pipenv-fork.readthedocs.io/en/latest/basics.html#example-pipenv-upgrade-workflow

### Checklist
Before creating a PR, run through this checklist and mark each as complete.

- [ ] I have written my commit messages in the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) format. 
- [ ] I have read the [CONTRIBUTING](/CONTRIBUTING.md) doc
- [ ] I have added tests (when possible) that prove my fix is effective or that my feature works
- [ ] I have checked that all unit tests pass after adding my changes
- [ ] I have updated necessary documentation
- [ ] I have rebased my branch onto master
- [ ] I will ensure my PR is targeting the master branch and pulling from my branch from my own fork

### CODE REVIEW

File: bin/setup_venv.sh [hunk 1]
```diff
@@ -176,15 +176,14 @@ source "${VIRTUAL_ENV}/bin/activate"
...
-# Installs wheel package management, so that pulumi requirements install quickly
-pip3 install wheel
-
-# Get nodeenv version so that node can be installed before we install Python
-# dependencies because pulumi_eks depends on the presence of node.
-pip3 install "$(grep nodeenv "${script_dir}/../pulumi/python/requirements.txt")"
+# Install certain utility packages like `nodeenv` and `wheel` that aid
+# in the installation of other build tools and dependencies
+# required by the other python packages.
+PIPENV_VERBOSITY=-1 PIPENV_PIPFILE="${script_dir}/../pulumi/python/Pipfile" pipenv install --dev
```
Comment (4141done) [inline bin/setup_venv.sh:186]: **annotation:**
Here we are:

1. Silencing an informational message telling us that pipenv is going to use the current venv with `PIPENV_VERBOSITY=-1`
2. Pointing `pipenv` at our `Pipfile` which is not in the current dir with `PIPENV_PIPFILE`
3. Only installing `dev` depenencies which are currently `nodeenv` and `wheel`

File: bin/setup_venv.sh [hunk 2]
```diff
@@ -195,12 +194,15 @@ else
 fi
 
 # Install general package requirements
-pip3 install --requirement "${script_dir}/../pulumi/python/requirements.txt"
+PIPENV_VERBOSITY=-1 PIPENV_PIPFILE="${script_dir}/../pulumi/python/Pipfile" pipenv install
```
Comment (4141done) [inline bin/setup_venv.sh:197]: **annotation:**
Here we are:

1. Silencing an informational message telling us that pipenv is going to use the current venv with `PIPENV_VERBOSITY=-1`
2. Pointing `pipenv` at our `Pipfile` which is not in the current dir with `PIPENV_PIPFILE`
3. Installing all packages

File: bin/setup_venv.sh [hunk 3]
```diff
@@ -195,12 +194,15 @@ else
...
-pip3 install "${script_dir}/../pulumi/python/utility/kic-pulumi-utils" &&
-  rm -rf "${script_dir}/../pulumi/python/utility/kic-pulumi-utils/.eggs" \
-    "${script_dir}/../pulumi/python/utility/kic-pulumi-utils/build" \
-    "${script_dir}/../pulumi/python/utility/kic-pulumi-utils/kic_pulumi_utils.egg-info"
+pip3 install "${script_dir}/../pulumi/python/utility/kic-pulumi-utils"
+
+rm -rf "${script_dir}/../pulumi/python/utility/kic-pulumi-utils/.eggs" \
+  "${script_dir}/../pulumi/python/utility/kic-pulumi-utils/build" \
+  "${script_dir}/../pulumi/python/utility/kic-pulumi-utils/kic_pulumi_utils.egg-info"
+
```
Comment (4141done) [inline bin/setup_venv.sh:205]: **annotation:**
Removing the `&&` prevents an issue where if the pip install does not work the script would not fail.  The `rm` operation should not fail either way.

File: bin/setup_venv.sh [hunk 4]
```diff
@@ -265,20 +267,15 @@ else
...
 # Download Pulumi CLI tooling
-#
-# Added some error checking to handle a failure here since it can be a bit confusing to the user (or at least to me)
-# when this bombs out and you can't figure out why. Note that the logic is specifically looking for "~=" in the
-# requirements file. Other comparisons that are legal in terms of the requirements file WILL fail with this logic.
-#
+# Regular expression and sed command from https://superuser.com/a/363878
 echo "Downloading Pulumi CLI into virtual environment"
-PULUMI_VERSION="$(grep '^pulumi~=.*$' "${script_dir}/../pulumi/python/requirements.txt" | cut -d '=' -f2 || true)"
+PULUMI_VERSION="$(pip list | grep 'pulumi ' | sed -nre 's/^[^0-9]*(([0-9]+\.)*[0-9]+).*/\1/p')"
```
Comment (4141done) [inline bin/setup_venv.sh:273]: **annotation:**
Here we read the pulumi version from pip to get the version that was actually installed.

File: bin/setup_venv.sh [hunk 5]
```diff
@@ -176,15 +176,14 @@ source "${VIRTUAL_ENV}/bin/activate"
 
 set -o nounset # abort on unbound variable
 
-# Use the latest version of pip
+# Use the latest version of pip and pipenv
 pip3 install --upgrade pip
+pip install pipenv
```
Comment (dekobon) [inline bin/setup_venv.sh:181]: Perhaps use `pip3` here to bring it in harmony with the above line.

**Review by dekobon — APPROVED** [review]
This is a great change! :fish_cake:

**Comment by qdzlug** [comment]
@4141done - this looks good to me; out of an abundance of caution I just kicked off all the deploy tests, but once they are clear I'm happy to have this merged whenever you are ready (my PR for #152 only touches the Jenkinsfile so it's going to be fine with yours).

**Review by qdzlug — APPROVED** [review]
Looks good to me...

**Pull request merged**